primitives/storage/src/storage_entry.rs

```diff
@@ -0,0 +1,40 @@
+use crate::{error::Error, StorageProofChecker};
+use codec::{Decode, Encode};
+use frame_support::ensure;
+use sp_runtime::traits::Header as HeaderT;
+use sp_std::prelude::Vec;
+
+#[derive(Default, Clone, Encode, Decode)]
+pub struct StorageEntry<V> {
+	pub key: Vec<u8>,
+	pub value: Option<V>,
+	pub proof: Option<Vec<Vec<u8>>>,
+}
+
+pub trait VerifyStorageProof {
+	fn verify_storage_proof<Header: HeaderT>(&self, header: &Header) -> Result<(), Error>;
+}
+
+impl VerifyStorageProof for StorageEntry<Vec<u8>> {
+	fn verify_storage_proof<Header: HeaderT>(&self, header: &Header) -> Result<(), Error> {
+		let proof = self.proof.as_ref().ok_or(Error::NoProofSupplied)?;
+		let actual = StorageProofChecker::<<Header as HeaderT>::Hashing>::check_proof(
+			*header.state_root(),
+			&self.key,
+			proof.to_vec(),
+		)?;
+
+		// Todo: Why do they do it like that, we could supply the proof only and get the value from the proof directly??
+		ensure!(actual == self.value, Error::WrongValue);
+		Ok(())
+	}
+}
+
+pub fn into_storage_entry_iter<'a, S>(
+	source: impl IntoIterator<Item = S> + 'a,
+) -> impl Iterator<Item = StorageEntry<Vec<u8>>> + 'a
+where
+	S: Into<StorageEntry<Vec<u8>>>,
+{
+	source.into_iter().map(|s| s.into())
+}
```

Comment:
Seems like a lot of code to essentially wrap a `.map()` 😅 but then again, it has a good name I think and is readable, so it's fine by me 👍

Reply:
Haha, you are right, before the latest commit where I extracted another function, this was called multiple times at places where we already had a lot of chaining.

But now, I am not certain if this is worth it. 😄